src/resources/uart_esp32.cc

```diff
@@ -17,95 +17,584 @@
+  uint32_t baud_rate;
+  uart_toit_hal_get_baudrate(hal_, &baud_rate);
+  return baud_rate;
+}
+
+void UartResource::set_tx_pin(gpio_num_t tx_pin) {
+  if (!try_set_iomux_pin(tx_pin, SOC_UART_TX_PIN_IDX)) {
+    gpio_hal_iomux_func_sel(GPIO_PIN_MUX_REG[tx_pin], PIN_FUNC_GPIO);
+    gpio_set_level(tx_pin, 1);
+    esp_rom_gpio_connect_out_signal(tx_pin, UART_PERIPH_SIGNAL(port_, SOC_UART_TX_PIN_IDX), false, false);
+  }
+}
+
+void UartResource::set_rx_pin(gpio_num_t rx_pin) {
+  if (!try_set_iomux_pin(rx_pin, SOC_UART_RX_PIN_IDX)) {
+    gpio_hal_iomux_func_sel(GPIO_PIN_MUX_REG[rx_pin], PIN_FUNC_GPIO);
+    gpio_set_pull_mode(rx_pin, GPIO_PULLUP_ONLY);
+    gpio_set_direction(rx_pin, GPIO_MODE_INPUT);
+    esp_rom_gpio_connect_in_signal(rx_pin, UART_PERIPH_SIGNAL(port_, SOC_UART_RX_PIN_IDX), false);
+  }
+}
+
+void UartResource::set_rts_pin(gpio_num_t rts_pin) {
+  if (!try_set_iomux_pin(rts_pin, SOC_UART_RTS_PIN_IDX)) {
+    gpio_hal_iomux_func_sel(GPIO_PIN_MUX_REG[rts_pin], PIN_FUNC_GPIO);
+    gpio_set_direction(rts_pin, GPIO_MODE_OUTPUT);
+    esp_rom_gpio_connect_out_signal(rts_pin, UART_PERIPH_SIGNAL(port_, SOC_UART_RTS_PIN_IDX), false, false);
+  }
+}
+
+void UartResource::set_cts_pin(gpio_num_t cts_pin) {
+  if (!try_set_iomux_pin(cts_pin, SOC_UART_CTS_PIN_IDX)) {
+    gpio_hal_iomux_func_sel(GPIO_PIN_MUX_REG[cts_pin], PIN_FUNC_GPIO);
+    gpio_set_pull_mode(cts_pin, GPIO_PULLUP_ONLY);
+    gpio_set_direction(cts_pin, GPIO_MODE_INPUT);
+    esp_rom_gpio_connect_in_signal(cts_pin, UART_PERIPH_SIGNAL(port_, SOC_UART_CTS_PIN_IDX), false);
+  }
+}
+
+UART_ISR_INLINE void UartResource::enable_interrupt_index_isr(uart_toit_interrupt_index_t index) {
+  IsrSpinLocker locker(&spinlock_);
+  enable_interrupt_mask_(interrupt_mask(index));
+}
+
+void UartResource::enable_interrupt_index(uart_toit_interrupt_index_t index) {
+  SpinLocker locker(&spinlock_);
+  enable_interrupt_mask_(interrupt_mask(index));
+}
+
+UART_ISR_INLINE void UartResource::disable_interrupt_index_isr(uart_toit_interrupt_index_t index) {
+  IsrSpinLocker locker(&spinlock_);
+  disable_interrupt_mask_(interrupt_mask(index));
+}
+
+UART_ISR_INLINE void UartResource::disable_read_interrupts() {
+  disable_interrupt_index_isr(UART_TOIT_INTR_RXFIFO_FULL);
+  disable_interrupt_index_isr(UART_TOIT_INTR_RX_TIMEOUT);
+}
+
+void UartResource::enable_read_interrupts() {
+  enable_interrupt_index(UART_TOIT_INTR_RXFIFO_FULL);
+  enable_interrupt_index(UART_TOIT_INTR_RX_TIMEOUT);
+}
+
+void UartResource::disable_interrupt_index(uart_toit_interrupt_index_t index) {
+  SpinLocker locker(&spinlock_);
+  disable_interrupt_mask_(interrupt_mask(index));
+}
+
+void UartResource::clear_interrupt_index(uart_toit_interrupt_index_t index) {
+  uart_toit_hal_clr_intsts_mask(hal_, interrupt_mask(index));
+}
+
+void UartResource::clear_interrupt_mask(uint32_t mask) {
+  uart_toit_hal_clr_intsts_mask(hal_, mask);
+}
+
+bool UartResource::try_set_iomux_pin(gpio_num_t pin, uint32_t iomux_index) const {
```

Comment:
What is this for, and what does the return value mean?

Reply:
I tried to add an appropriate comment. But it basicaly tries to connect the pin via the direct IO mux (for higher speed). If that does not work, the individual pin configurations needs to go via the GPIO matrix (slower speeds). For UART it probably does not matter at all. But the driver in the esp-idf does it.